parameter_expansion/tests/test_pe.py

```diff
@@ -1,212 +1,609 @@
             aa="FOO",
             bb="BAR",
         ),
-        "-aa/bb/cc/FOO/BAR-",
-    ),
-}
-
-test_envs = (
-    {
-        "parameter": "set",
-        "word": "word",
-    },  # set and not null
-    {
-        "parameter": "",
-        "word": "word",
-    },  # set but null
-    {
-        "word": "word",
-    },  # unset
-)
-
-test_case_map = (
-    "Set and not null",
-    "Set, but null",
-    "Unset",
-)
-
-
-def test():
-    for string, tc in subst_test_cases.items():
-        for i, env in enumerate(test_envs):
-            env = dict(env)  # Don't allow tests to mutate each other's envs
-            try:
-                result = string, pe.expand(string, env), tc[i], test_case_map[i]
-                assert result[1] == result[2], result
-            except pe.ParameterExpansionNullError:
-                assert tc[i] == "error", (string, tc[i], test_case_map[i])
-
-    for string, (parameter, expected) in affix_test_cases.items():
-        env = {"parameter": parameter}
-        assert pe.expand(string, env) == expected
-
-
-def test_substring():
-    for string, (parameter, expected) in substring_test_cases.items():
-        env = {"parameter": parameter}
-        assert pe.expand(string, env) == expected, (string, parameter)
-
-
-def test_replace():
-    for string, (parameter, expected) in replace_test_cases.items():
-        env = {"parameter": parameter}
-        assert pe.expand(string, env) == expected, (string, parameter)
-
-
-def test_simple():
-    for string, (env, expected) in simple_test_cases.items():
-        assert pe.expand(string, env) == expected, (string, env)
+    ),
+]
+
+
+@pytest.mark.parametrize("test", subst_test_cases)
+def test_expand(test):
+    try:
+        result = pex.expand(test.tested_shell, env=test.env)
+        assert result == test.expected_str
+    except pex.ParameterExpansionNullError:
+        assert test.expected_str == "error"
+
+
+@pytest.mark.parametrize("test", affix_test_cases)
+def test_affix(test):
+    assert pex.expand(test.tested_shell, env=test.env) == test.expected_str
+
+
+@pytest.mark.parametrize("test", substring_test_cases)
+def test_substring(test):
+    assert pex.expand(test.tested_shell, env=test.env) == test.expected_str
+
+
+@pytest.mark.parametrize("test", replace_test_cases)
+def test_replace(test):
+    assert pex.expand(test.tested_shell, env=test.env) == test.expected_str
+
+
+@pytest.mark.parametrize("test", simple_test_cases)
+def test_simple(test):
+    assert pex.expand(test.tested_shell, env=test.env) == test.expected_str
+
+
+@pytest.mark.parametrize("test", simple_simple_test_cases)
+def test_expand_simple(test):
+    assert pex.expand(test.tested_shell, env=test.env) == test.expected_str
+
+
+def test_expand_strict_raises_Exception():
+    string = "- $parameter/$aa/${bb}"
+    env = dict(foo="bar", bar="baz")
+    try:
+        pex.expand(string, env, strict=True)
+        raise Exception("ParameterExpansionNullError should be raised")
```

Comment:
Please raise an AssertionError here instead of a bare exception.

Reply:
Fixed.